sshd/src/test/scala/ammonite/repl/sshd/SshTestingTraits.scala

```diff
@@ -0,0 +1,131 @@
+    login ← nonEmptyAlphaNumStr
+    password ← alphaNumStr
+  } yield (login, password)
+
+  def genCredsPair = for {
+    creds1 ← genCreds
+    creds2 ← genCreds
+  } yield (creds1, creds2)
+
+  def genNonMatchingCredsPair = genCredsPair.suchThat(pair ⇒ pair._1 != pair._2)
+
+  def uniqueUsersWithOneOfThem = for {
+    users:List[(String, String)] ← Gen.nonEmptyListOf(genCreds)
+    noDuplicatedLogins = users.groupBy(_._1).values.map(_.head).toList
+    someUserIndex ← Gen.chooseNum(0, noDuplicatedLogins.size - 1)
+  } yield (noDuplicatedLogins, noDuplicatedLogins(someUserIndex))
+
+  def withTmpDirectory[T](block:Path ⇒ T):T = {
+    lazy val tmpDir = Path(Path.makeTmp)
+    try {
+      block(tmpDir)
+    } finally {
+      rm! tmpDir
+    }
+  }
+
+  def testSshServer(user:(String, String), shell: SshServer.RemoteShellServer)(implicit dir:Path) = {
+    SshServer(SshServerConfig(port = 0, username = user._1, password = user._2, ammoniteHome = dir), shell)
+  }
+
+  def withTestSshServer[T](user:(String, String), shell: SshServer.RemoteShellServer)
+                          (test:org.apache.sshd.SshServer ⇒ T)
+                          (implicit dir:Path):T = {
+    val server = testSshServer(user, shell)
+    server.start()
+    try {
+      test(server)
+    } finally {
+      server.stop()
+    }
+  }
+
+  def sshClient(creds: (String, String), sshServer: org.apache.sshd.SshServer):Session = {
+    sshClient(creds, Option(sshServer.getHost).getOrElse("localhost"), sshServer.getPort)
+  }
+
+  def sshClient(creds: (String, String), host:String, port:Int):Session = {
+    val client = new JSch
+    val session = client.getSession(creds._1, host, port)
+    session.setPassword(creds._2)
+    session.setUserInfo(autoAcceptHostKeyUser(creds._2))
+    session
+  }
+
+  private def autoAcceptHostKeyUser(password:String) = new UserInfo {
+    override def promptPassword(message: String): Boolean = true
+    override def promptPassphrase(message: String): Boolean = false
+    override def promptYesNo(message: String): Boolean = true
+
+    override def showMessage(message: String): Unit = { /*noop*/ }
+    override def getPassword: String = password
+    override def getPassphrase: String = ???
+  }
+
+
+  class Shell private(shell:Channel) {
+    def this(sshClient:Session) = this(sshClient.openChannel("shell"))
+
+    val input = {
+      val pipeEnd = new PipedInputStream(2048)
+      val pipeStart = new PipedOutputStream(pipeEnd)
+      shell.setInputStream(pipeEnd)
+
+      new PrintStream(pipeStart, /*autoflush = */true)
+    }
+    val output:LineReader = {
+      val pipeEnd = new PipedInputStream(2048)
+      val pipeStart = new PipedOutputStream(pipeEnd)
+      shell.setOutputStream(pipeStart)
+
+      new InputStreamReader(pipeEnd) with LineReader
+    }
+
+    def connect() = shell.connect()
+    def isConnected = shell.isConnected
+    def disconnect() = shell.disconnect()
+  }
+
+
+  trait LineReader { reader:Reader ⇒
```

Comment:
I only see this being used once. AFAICT this should just be a helper method on a static object